services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowDiversitySpec.groovy

```diff
@@ -0,0 +1,326 @@
+                    isl.dstSwitch.dpId.toString(), isl.dstPort, ["cost": (diversityIslWeight + 1).toString()])])
+        }
+
+        when: "Create the second flow with diversity enabled"
+        def flow2 = flowHelper.randomFlow(srcSwitch, dstSwitch, false, [flow1]).tap { it.diverseFlowId = flow1.id }
+        flowHelper.addFlow(flow2)
+        def flow2Path = PathHelper.convert(northbound.getFlowPath(flow2.id))
+
+        then: "The flow is built through the most preferable path (path of the first flow)"
+        flow2Path == flow1Path
+
+        when: "Create the third flow with diversity enabled"
+        def flow3 = flowHelper.randomFlow(srcSwitch, dstSwitch, false, [flow1, flow2]).tap {
+            it.diverseFlowId = flow2.id
+        }
+        flowHelper.addFlow(flow3)
+        def flow3Path = PathHelper.convert(northbound.getFlowPath(flow3.id))
+
+        then: "The flow is built through one of the alternative paths because they are preferable already"
+        def involvedIsls = [flow2Path, flow3Path].collect { pathHelper.getInvolvedIsls(it) }.flatten()
+        flow3Path != flow2Path
+        involvedIsls.unique(false) == involvedIsls
+
+        and: "Delete flows and link props"
+        [flow1, flow2, flow3].each { flowHelper.deleteFlow(it.id) }
+        northbound.deleteLinkProps(northbound.getAllLinkProps())
+    }
+
+    def "Able to get flow paths with correct overlapping segments stats for diverse flows"() {
+        given: "Two active neighboring switches with three not overlapping paths at least"
+        def switches = topology.getActiveSwitches()
+        def (Switch srcSwitch, Switch dstSwitch) = [switches, switches].combinations()
+                .findAll { src, dst -> src != dst }.find { Switch src, Switch dst ->
+            database.getPaths(src.dpId, dst.dpId)*.path.collect {
+                pathHelper.getInvolvedIsls(it)
+            }.unique { a, b -> a.intersect(b) ? 0 : 1 }.size() >= 3
+
+        } ?: assumeTrue("No suiting switches found", false)
+
+        and: "Create three flows with diversity enabled"
+        def flow1 = flowHelper.randomFlow(srcSwitch, dstSwitch, false)
+        def flow2 = flowHelper.randomFlow(srcSwitch, dstSwitch, false, [flow1]).tap { it.diverseFlowId = flow1.id }
+        def flow3 = flowHelper.randomFlow(srcSwitch, dstSwitch, false, [flow1, flow2]).tap {
+            it.diverseFlowId = flow2.id
+        }
+        [flow1, flow2, flow3].each { flowHelper.addFlow(it) }
+
+        def (flow1Path, flow2Path, flow3Path) = [flow1, flow2, flow3].collect { northbound.getFlowPath(it.id) }
+        def allInvolvedIsls = [flow1Path, flow2Path, flow3Path].collect {
+            pathHelper.getInvolvedIsls(PathHelper.convert(it))
+        }.flatten()
+        assert allInvolvedIsls.unique(false) == allInvolvedIsls
+
+        expect: "Flow path response for all flows has correct overlapping segments stats"
+        def flow2SwitchCount = pathHelper.getInvolvedSwitches(flow2.id).size()
```

Comment:
since you already have paths for each flow it is better to call `getInvolvedSwitches(path)` instead of passing the flow id